components/ide/jetbrains/gateway-plugin/src/main/kotlin/io/gitpod/jetbrains/gateway/GitpodConnectionProvider.kt

```diff
@@ -234,18 +259,27 @@ class GitpodConnectionProvider : GatewayConnectionProvider {
                                     )
                                     val client = connector.connect()
                                     client.clientClosed.advise(connectionLifetime) {
+                                        thisLogger().d("clientClosed, ${client.prettyPrint()}")
                                         application.invokeLater {
+                                            connectionLifetime.onTermination {
+                                                thisLogger().d("clientTermination, clientClosed didChange fired, ${client.prettyPrint()}")
+                                                dispatcher.multicaster.didChange()
+                                            }
                                             connectionLifetime.terminate()
                                         }
                                     }
                                     client.onClientPresenceChanged.advise(connectionLifetime) {
+                                        thisLogger().d("presenceChanged, ${client.prettyPrint()}")
                                         application.invokeLater {
+                                            thisLogger().d("clientPresenceChanged didChange fired, ${client.prettyPrint()}")
+                                            dispatcher.multicaster.didChange()
                                             if (client.clientPresent) {
                                                 statusMessage.text = ""
                                             }
                                         }
                                     }
                                     thinClient = client
+                                    jetbrainsClientMap[update.workspaceId] = client
```

Comment:
Should not it be like:
```java
jetbrainsClientMap[update.workspaceId] = client
dispatcher.multicaster.didChange()
connectionLifetime.onTermination {
    jetbrainsClientMap.remove(update.workspaceId)
    dispatcher.multicaster.didChange()
}
```

We could also create a private function `setCleint(id, client | null)` and then do here:
```java
setClient(id, client);
connectionLifetime.onTermination {
    setClient(id, null) 
}
```